File: Python/1232.CheckIfItIsAStraightLine.py
# Given list of (x, y) coordinates checkStraightLine checks if it forms straight line
# 
class Solution:
    def checkStraightLine(self, coordinates: List[List[int]]) -> bool:
        if len(coordinates) <= 2:
            # it is always possible to create line through 2 points (and also 1 and 0)
            return True

        x0, y0 = coordinates[0]
        x1, y1 = coordinates[1]

        if x0 == x1:
            # check if this is vertical line (all x are the same)
            return all([x == x0 for x, _ in coordinates])

        # 2 points define line, so find it's equation y = a*x + b
        a = (y0 - y1) / (x0 - x1)
        b = y0 - a * x0

        line = lambda x: a*x + b

        #check if all points lies on the line defined by (x0, y0) and (x1, y1)
        return all([y == line(x) for x, y in coordinates])

File: C++/1232.CheckIfItIsAStraight Line.cpp
class Solution {
public:
    bool checkStraightLine(vector<vector<int>>& coordinates) {
        int n=coordinates.size();

        if(n==2)
            return true;
        /*
           A straight line has one slope! If any 2 points on a straight line have a slope equal to the
           slope of the given line, they form a straight line.

           Slope of (x1, y1) and (x2, y2) = (y2-y1)/(x2-x1)
           If the slope of (xi, yi) is equal to the above slope, return true, else return false.
        */
        int x1=coordinates[1][0];
        int y1=coordinates[1][1];
        int x2=coordinates[0][0];
        int y2=coordinates[0][1];

        int y_diff = y2-y1;
        int x_diff = x2-x1;

        for(int i=2; i<n; i++)
        {
            int x = coordinates[i][0] - x2;
            int y = coordinates[i][1] - y2;

            // m1(y_diff/x_diff) = m2(y/x), checking the slopes
            if(x * y_diff != x_diff * y)
                return false;
        }
        return true;
    }
};
